microsoft/PowerToys
# Improving performance of Adding and Removing Packaged Apps

## Summary of the Pull Request
The previous implementation (taken from wox) provided the ability to manually re-index uwp and win32 programs via a settings button.  Because there was currently no settings button, a temporary workaround was put place in utilizing a file watcher in program files.    This was problematic in that certain applications such as MongoDB and Postgres continually write to the ProgramFiles directory, which would trigger a reindex of the UWP packages.

This is UWP part of the implementation that leverages the PacakageCatalog to be notified when packaged applications are installed and/removed.   Win32 applications will be handled separately. 

Win32 applications will be handled in a separate PR. 

 
## References


## PR Checklist
* [x] Applies to #3208


## Detailed Description of the Pull Request / Additional comments
1. Encapsulates storage and management of the list of UWP apps in a new class PackageRepository. 
2. Allows for Insert / Delete when new applications are detected, removing the need to 


## Validation Steps Performed
1. Validated that installed/uninstalled applications can be detected on the fly. 
1.1 Types 'terminal in the search bar', verified that the Packaged application wasn't in the search results 
1.2 Installed Terminal from the store.  
1.3 Resubmitted the terminal query, and verified that it appeared in the results. 
1.4 Uninstalled Terminal via the system settings. 
1.5 Verified that terminal didn't appear in the results. 
2. Validated that cached information is still able to be deserialized into the Uwp Applications. 
2.1 Installed Powertoys 18. 
2.2.  Sign out of windows to force a save to the cache. 
2.3. Validate that that UWP.cache was written to in `AppData\Local\Microsoft\PowerToys\PowerToys Run\Cache`
2.4 Ran from the debugger and validated that the UWP.cache was properly deserialized. 
3. Validated theme changes are picked up. 
3.1 Switched to light theme.  Verified breakpoints are hit.  Observed Packaged Edge w/ black icon
3.2. Switched to dark theme. Verified breakpoints are hit.  Observed Edge with white icon.
